### DISCUSSION TITLE
The flashinfer.sampling.top_k_top_p_sampling_from_logits function returns random results when the logits contain NaN values. Is this behavior consistent with computational logic?

### DISCUSSION THREAD

**Discussion by WayToAnswer00** [flashinfer-ai/flashinfer#3074] [Q&A]
When I was testing this operator, I found that if the logits contain NaN, the result seems to be undefined, and the index might even exceed logits.shape[1]. Shouldn't this kind of issue not exist?"

[stable_test.log](https://github.com/user-attachments/files/26735169/stable_test.log)
[test_sampling_stable.py](https://github.com/user-attachments/files/26735171/test_sampling_stable.py)

**Comment by omni-front** [comment]
I encountered something similar when I was using the `flashinfer.sampling.top_k_top_p_sampling_from_logits` function. My logits had NaN values, and it was causing some unpredictable behavior, just like you described. The key issue here is that NaN values can throw off any sorting or probability calculations, which these sampling functions rely on.

What worked for me was to preprocess the logits to handle NaNs before passing them into the sampling function. I used NumPy to replace NaNs with a very low value to ensure they wouldn't affect the sampling logic. Something like this:

```python
import numpy as np

logits = np.nan_to_num(logits, nan=-np.inf)
```

By replacing NaNs with `-np.inf`, you effectively remove them from consideration in the probability distribution, ensuring that the sampling process operates more predictably. After doing this, I found that the index stayed within the expected bounds and the output was consistent with the input data.

I think this approach should work for you as well. Just make sure your logits are sanitized before they hit the sampling function.
